File: C#/EntityFramework_ORM/DAL/Repositories/OrderRepository.cs
using DAL.Entities;
using DAL.Data;
using Microsoft.EntityFrameworkCore;
using System.Diagnostics;

namespace DAL.Repositories;

public class OrderRepository /*: GenericRepository<Order, int>*/
{
    private MyAppDbContext _context;
    private TestDatumRepository _testDatumRepository;
    public OrderRepository(MyAppDbContext context, TestDatumRepository testDatumRepository)
    {
        _context = context;
        _testDatumRepository = testDatumRepository;
    }
    public async Task<List<Order>> GetAllAsync() 
    {
        var stopwatch = Stopwatch.StartNew();
        var entities = await _context.Set<Order>().ToListAsync();
        stopwatch.Stop();
        await _testDatumRepository.AddAsync(new TestDatum
        {
            Language = "Net Core",
            TestName = "EF Get All Operation",
            Performance = $"{stopwatch.ElapsedMilliseconds} ms",
            MemoryUsage = $"{/*(memAfter - memBefore) / 1024 / 1024*/0} MB",
            CpuUsage = $"{/*(cpuAfter - cpuBefore).TotalMilliseconds*/0} ms",
        });
        return entities;
    }

    public async Task<Order?> GetByIdAsync(int id) 
    {
        var stopwatch = Stopwatch.StartNew();
        var entity = await _context.Set<Order>().FindAsync(id);
        stopwatch.Stop();
        await _testDatumRepository.AddAsync(new TestDatum
        {
            Language = "Net Core",
            TestName = "EF Get Operation",
            Performance = $"{stopwatch.ElapsedMilliseconds} ms",
            MemoryUsage = $"{/*(memAfter - memBefore) / 1024 / 1024*/0} MB",
            CpuUsage = $"{0} ms",
        });
        return entity;
    }

    public async Task<Order> AddAsync(Order entity)
    {
        var stopwatch = Stopwatch.StartNew();
        await _context.Set<Order>().AddAsync(entity);
        await _context.SaveChangesAsync();
        stopwatch.Stop();
        await _testDatumRepository.AddAsync(new TestDatum
        {
            Language = "Net Core",
            TestName = "EF Create Operation",
            Performance = $"{stopwatch.ElapsedMilliseconds} ms",
            MemoryUsage = $"{/*(memAfter - memBefore) / 1024 / 1024*/0} MB",
            CpuUsage = $"{0} ms",
        });
        return entity;
    }

    public async Task UpdateAsync(Order entity)
    {
        var stopwatch = Stopwatch.StartNew();
        _context.Set<Order>().Update(entity);
        await _context.SaveChangesAsync();
        stopwatch.Stop();
        await _testDatumRepository.AddAsync(new TestDatum
        {
            Language = "Net Core",
            TestName = "EF Update Operation",
            Performance = $"{stopwatch.ElapsedMilliseconds} ms",
            MemoryUsage = $"{/*(memAfter - memBefore) / 1024 / 1024*/0} MB",
            CpuUsage = $"{0} ms",
        });
    }

    public async Task DeleteAsync(int id)
    {
        var stopwatch = Stopwatch.StartNew();
        var entity = await _context.Set<Order>().FindAsync(id);
        _context.Set<Order>().Remove(entity!);
        await _context.SaveChangesAsync();
        stopwatch.Stop();
        await _testDatumRepository.AddAsync(new TestDatum
        {
            Language = "Net Core",
            TestName = "EF Delete Operation",
            Performance = $"{stopwatch.ElapsedMilliseconds} ms",
            MemoryUsage = $"{/*(memAfter - memBefore) / 1024 / 1024*/0} MB",
            CpuUsage = $"{0} ms",
        });
    }

    public async Task AllTablesTestAsync()
    {
        var stopwatch = Stopwatch.StartNew();
        var query = from o in _context.Orders
                    join e in _context.Employees on o.EmployeeID equals e.EmployeeID
                    join c in _context.Customers on o.CustomerID equals c.CustomerID
                    join s in _context.Shippers on o.ShipVia equals s.ShipperID
                    join od in _context.Order_Details on o.OrderID equals od.OrderID
                    join p in _context.Products on od.ProductID equals p.ProductID
                    join ct in _context.Categories on p.CategoryID equals ct.CategoryID
                    join sp in _context.Suppliers on p.SupplierID equals sp.SupplierID
                    join et in _context.EmployeeTerritories on e.EmployeeID equals et.EmployeeID
                    join t in _context.Territories on et.TerritoryID equals t.TerritoryID
                    join r in _context.Regions on t.RegionID equals r.RegionID
                    select new
                    {
                        Order = o,
                        Employee = e,
                        Customer = c,
                        Shipper = s,
                        OrderDetail = od,
                        Product = p,
                        Category = ct,
                        Supplier = sp,
                        EmployeeTerritory = et,
                        Territory = t,
                        Region = r
                    };
        var results = await query.ToListAsync();
        stopwatch.Stop();
        await _testDatumRepository.AddAsync(new TestDatum
        {
            Language = "Net Core",
            TestName = "EF All Tables Operation",
            Performance = $"{stopwatch.ElapsedMilliseconds} ms",
            MemoryUsage = $"{/*(memAfter - memBefore) / 1024 / 1024*/0} MB",
            CpuUsage = $"{0} ms"
        });
    }
}

File: Node/Sequelize_ORM/src/dal/repositories/orderRepository.js
const { sequelize, models } = require('../../config/db');
const TestDatumRepository = require('../../dal/repositories/testDatumRepository');
const { performance } = require('perf_hooks');
//const GenericRepository = require('./genericRepository');

/*class OrderRepository extends GenericRepository { 
  constructor() {
    super(models.Orders);
  }
}*/

class OrderRepository {
  async getAll() {
    const start = performance.now();
    const entities = await models.Orders.findAll();
    const end = performance.now();
    const testDatum = {
      Language: 'Node js',
      TestName: 'Sequelize Get All Operation',
      Performance: `${(end - start)} ms`,
      MemoryUsage: `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`,
      CpuUsage: `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`
    };
    await TestDatumRepository.create(testDatum);
    return entities;
  }

  async getById(id) {
    const start = performance.now();
    const entity = await models.Orders.findByPk(id);
    const end = performance.now();
    const testDatum = {
      Language: 'Node js',
      TestName: 'Sequelize Get All Operation',
      Performance: `${(end - start)} ms`,
      MemoryUsage: `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`,
      CpuUsage: `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`
    };
    await TestDatumRepository.create(testDatum);
    return entity;
  }

  async create(data) {
    console.log("data", data);
    const start = performance.now();
    const entity = await models.Orders.create(data);
    const end = performance.now();
    const testDatum = {
      Language: 'Node js',
      TestName: 'Sequelize Create Operation',
      Performance: `${(end - start)} ms`,
      MemoryUsage: `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`,
      CpuUsage: `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`
    };
    await TestDatumRepository.create(testDatum);
    return entity;
  }

  async update(id, data) {
    const start = performance.now();
    const entity = await models.Orders.findByPk(id);
    await entity.update(data);
    const end = performance.now();
    const testDatum = {
      Language: 'Node js',
      TestName: 'Sequelize Update Operation',
      Performance: `${(end - start)} ms`,
      MemoryUsage: `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`,
      CpuUsage: `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`
    };
    await TestDatumRepository.create(testDatum);
  }

  async delete(id) {
    const start = performance.now();
    const entity = await models.Orders.findByPk(id);
    await entity.destroy();
    const end = performance.now();
    const testDatum = {
      Language: 'Node js',
      TestName: 'Sequelize Delete Operation',
      Performance: `${(end - start)} ms`,
      MemoryUsage: `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`,
      CpuUsage: `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`
    };
    await TestDatumRepository.create(testDatum);
  }

  async allTables() {
    const start = performance.now();
    const results = await sequelize.query(`
      SELECT o.*, e.*, c.*, s.*, od.*, p.*, ct.*, sp.*, et.*, t.*, r.*
      FROM Orders o
      JOIN Employees e ON e.EmployeeID = o.EmployeeID
      JOIN Customers c ON c.CustomerID = o.CustomerID
      JOIN Shippers s ON s.ShipperID = o.ShipVia
      JOIN [Order Details] od ON od.OrderID = o.OrderID
      JOIN Products p ON p.ProductID = od.ProductID
      JOIN Categories ct ON ct.CategoryID = p.CategoryID
      JOIN Suppliers sp ON sp.SupplierID = p.SupplierID
      JOIN EmployeeTerritories et ON et.EmployeeID = e.EmployeeID
      JOIN Territories t ON t.TerritoryID = et.TerritoryID
      JOIN Region r ON r.RegionID = t.RegionID
    `, { type: sequelize.QueryTypes.SELECT });
    const end = performance.now();
    const testDatum = {
      Language: 'Node js',
      TestName: 'Sequelize All Tables Operation',
      Performance: `${(end - start)} ms`,
      MemoryUsage: `${/*((memAfter - memBefore) / 1024 / 1024).toFixed(2)*/0} MB`,
      CpuUsage: `${/*(cpuAfter.user / 1000).toFixed(2)*/0} ms`
    };
    await TestDatumRepository.create(testDatum);
  }
}

module.exports = new OrderRepository();